gtsfm/frontend/matcher/cacher/matcher_cacher.py

```diff
@@ -0,0 +1,181 @@
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+        match_indices: np.ndarray,
+    ) -> None:
+        """Save the results to the cache."""
+        cache_path = self.__get_cache_path(
+            cache_key=self.__generate_cache_key(
+                keypoints_i1=keypoints_i1,
+                keypoints_i2=keypoints_i2,
+                descriptors_i1=descriptors_i1,
+                descriptors_i2=descriptors_i2,
+                im_shape_i1=im_shape_i1,
+                im_shape_i2=im_shape_i2,
+            )
+        )
+        io_utils.write_to_compressed_file(match_indices, cache_path)
+
+    def match(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+    ) -> np.ndarray:
+        """Match descriptor vectors.
+
+        If the results are in the cache, they are fetched and returned. Otherwise, the `detect_and_describe` of the
+        underlying object's API is called and the results are cached.
+
+        # Some matcher implementations (such as SuperGlue) utilize keypoint coordinates as
+        # positional encoding, so our matcher API provides them for optional use.
+
+        Output format:
+        1. Each row represents a match.
+        2. First column represents keypoint index from image #i1.
+        3. Second column represents keypoint index from image #i2.
+        4. Matches are sorted in descending order of the confidence (score), if possible.
+
+        Args:
+            keypoints_i1: keypoints for image #i1, of length N1.
+            keypoints_i2: keypoints for image #i2, of length N2.
+            descriptors_i1: descriptors corr. to keypoints_i1.
+            descriptors_i2: descriptors corr. to keypoints_i2.
+            im_shape_i1: shape of image #i1, as (height,width).
+            im_shape_i2: shape of image #i2, as (height,width).
+
+
+        Returns:
+            Match indices (sorted by confidence), as matrix of shape (N, 2), where N < min(N1, N2).
+        """
+        cached_data = self.__load_result_from_cache(
+            keypoints_i1=keypoints_i1,
+            keypoints_i2=keypoints_i2,
+            descriptors_i1=descriptors_i1,
+            descriptors_i2=descriptors_i2,
+            im_shape_i1=im_shape_i1,
+            im_shape_i2=im_shape_i2,
+        )
+
+        if cached_data is not None:
+            return cached_data
+
+        match_indices = self._matcher.match(
+            keypoints_i1=keypoints_i1,
+            keypoints_i2=keypoints_i2,
+            descriptors_i1=descriptors_i1,
+            descriptors_i2=descriptors_i2,
+            im_shape_i1=im_shape_i1,
+            im_shape_i2=im_shape_i2,
+        )
+
+        self.__save_result_to_cache(
```

Comment:
what happens when we save to cache? doesn't push to your repo right? what's the need to save it here/ long-term strategy?

Reply:
No, this will just cache to the machine its being run on. It is meant for developers to iterate over multi-view optimizer quickly. 

When a dev wants to share the cached results with others (and kind of formalize the use of the dataset and its metrics), they should submit a PR to gtsfm-cache.

For example, if we decide to add skydio-501 to regular use, we will push its cache to the repo.